cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java

```diff
@@ -57,18 +66,36 @@
    */
   @Override
   public void create(NamespaceMeta namespaceMeta) throws IOException, ExploreException, SQLException {
-    Location namespaceHome = namespacedLocationFactory.get(namespaceMeta.getNamespaceId().toId());
-    if (namespaceHome.exists()) {
-      LOG.warn("Home directory '{}' for namespace '{}' already exists. Deleting it.",
-               namespaceHome, namespaceMeta.getName());
-      if (!namespaceHome.delete(true)) {
-        throw new IOException(String.format("Error while deleting home directory '%s' for namespace '%s'",
-                                            namespaceHome, namespaceMeta.getName()));
+    Location namespaceHome;
+    String rootDirectory = namespaceMeta.getConfig().getRootDirectory();
+    if (Strings.isNullOrEmpty(rootDirectory)) {
+      // no namespace custom location was provided one must be created by cdap
+      namespaceHome = namespacedLocationFactory.get(namespaceMeta.getNamespaceId().toId());
+      if (namespaceHome.exists()) {
+        throw new FileAlreadyExistsException(namespaceHome.toString());
+
+      }
+      // create namespace home dir
+      if (!namespaceHome.mkdirs()) {
+        throw new IOException(String.format("Error while creating home directory '%s' for namespace '%s'",
+                                            namespaceHome, namespaceMeta.getNamespaceId()));
+      }
+    } else {
+      // a custom location was provided
+      // check that its an absolute path
+      File customLocation = new File(rootDirectory);
+      if (!customLocation.isAbsolute()) {
+        throw new IOException(String.format("Cannot create the namespace '%s' with the given custom " +
+                                              "location %s. Custom location must be absolute path.",
+                                            namespaceMeta.getName(), rootDirectory));
+      }
+      // since this is a custom location we expect it to exists
+      if (!customLocation.exists()) {
+        throw new IOException(String.format("The provided home directory '%s' for namespace '%s' does not exists. " +
+                                              "Please create it on filesystem with sufficient privileges for the user" +
```

Comment:
good to capture the user in this message? I know its not part of your PR though, so let's add a TODO?